Nag

johanmalm opened the pull request:
--EDIT-- For after 0.8.4

For anyone wanting to test this (and assuming you're not installing it system wide), you need to create a symlink from build/clients/labnag to a location in `$PATH`.

The exit code relates to the position of the button as demonstrated in this example: ~~`labnag -m 'Foo' -Z 1 : -Z 2 : -Z 3 : -s 0`~~ (--EDITED-- See https://github.com/labwc/labwc/pull/2699#issuecomment-2993658830 for up-to-date example)

This PR does not deal with displaying config file error messages. Suggest we explore that in a separate PR - possibly by exploring something like `xmlCtxtSetErrorHandler()` or similar.

TODO:

- ~~In `action_prompt_create()` we ought to consider adding a config option
      for the command.~~ (--EDIT-- Do this at some other stage)
- [x] --EDIT-- Change examples to use `<prompt>` in `menu.xml`
- [x] --EDIT-- Update `labwc-config(5)`

johanmalm commented:
Example usage:

```xml
<keybind key="W-q">
  <action name="If">
    <prompt text="Quit?" option1="Yes" option2="No"/>
    <then><action name="Exit"/></then>
  </action>
</keybind>
```

tokyo4j commented:
Do we really need to support both `<prompt text="Quit?" option1="Yes" option2="No"/>` and `<prompt>Quit?</prompt>` syntaxes? I suggest supporting only the latter syntax and make the prompt command configurable like this:
```xml
<core>
   <promptCommand>labnag - %s -Z "No" : -s "Yes"</promptCommand>
</core>
```

Regarding @Consolatis's comment:
```c
		/*
		 * TODO: figure out how to have more than 2 choices,
		 *       maybe by using branch names like cond_0, cond_1
		 *       and so on and falling back to 'then' and 'else'
		 *       if not found?
		 */
```

I think we can set this out-of-scope. For those who want to do complex things like having multiple choices, they can write their own script and run `wtype` to trigger actions.

Also, I wonder if we should support the following options in `labnag`:
- `-B, --button-no-terminal`
- `-Z, --button-dismiss-no-terminal`
- `-s, --dismiss-button`

https://github.com/i3/i3/issues/2199 is the background for the first 2 options. In short, `i3-nagbar` spawns commands in terminal emulator (set with `$TERMINAL`) with `--button`/`--button-dismiss` options for historical reasons, and `*-no-terminal` options are added to allow launching commands without terminal emulator installed. And then swaybar supported them for compatibility.

Can we just not run commands in terminal emulator and drop those 2 options? IMO, we don't need to focus on compatibility with `swaybar`/`i3-nagbar` and we should prioritize simplicity to make options easier for starters.

Also for simplicity, I suggest dropping the last option. I was confused to find there are both `--dismiss-button` and `--button-dismiss`.

Consolatis commented:
> I think we can set this out-of-scope. For those who want to do complex things like having multiple choices, they can write their own script and run wtype to trigger actions.

Agreed.

Another issue with the prompt implementation I just noticed is that if a prompt is spawned and then a `Reconfigure` occurs we will execute free'd actions. So this need some reconfigure trigger to simply destroy all prompts, a button click on the GUI prompt would then not do anything but I think that's fine when somebody does a reconfigure in between spawning the prompt and clicking a button.

johanmalm commented:
Thanks for looking at this.

> Do we really need to support both `<prompt text="Quit?" option1="Yes" option2="No"/>` and `<prompt>Quit?</prompt>` syntaxes? I suggest supporting only the latter syntax

Agree with removing the options attributes.

Might it be worth using `<prompt text="foo"/>` in case we want to extend in future. If we use `<prompt>foo</prompt>` we cannot add attributes or child elements in future without breaking our 'clean' syntax.

> ...and make the prompt command configurable like this:
> 
> ```
> <core>
>    <promptCommand>labnag - %s -Z "No" : -s "Yes"</promptCommand>
> </core>
> ```

I like this in principle. Assume you mean `-m` instead of the lonely `-`?

By the way, have you explore the stdin functionality for future config-error-logging. Try `echo "foo\nbar" | labnag -l -m baz`


> Also, I wonder if we should support the following options in `labnag`:
>     * `-B, --button-no-terminal`
>     * `-Z, --button-dismiss-no-terminal`

Do you mean that we should support `Z` and `B` but remove support for `z` and `b`? I'm thinking it's better to just pass things to `execlp("sh", "sh", "-c", button->action, NULL);` rather than invoking terminals.

>     * `-s, --dismiss-button`

Do you mean to remove support for `-s` and simply use `-z foo :` instead?

> IMO, we don't need to focus on compatibility with `swaybar`/`i3-nagbar` and we should prioritize simplicity to make options easier for starters.

Agree.

Consolatis commented:
> Do we really need to support both `<prompt text="Quit?" option1="Yes" option2="No"/>` and `<prompt>Quit?</prompt>` syntaxes? I suggest supporting only the latter syntax and make the prompt command configurable like this:
> 
> ```
> <core>
>    <promptCommand>labnag - %s -Z "No" : -s "Yes"</promptCommand>
> </core>
> ```

One thing that doesn't work with this approach is having multiple actions with different option values.

tokyo4j commented:
> Might it be worth using <prompt text="foo"/> in case we want to extend in future. If we use <prompt>foo</prompt> we cannot add attributes or child elements in future without breaking our 'clean' syntax.

Agreed. If we decide to put `<prompt>` in `<action name="If">`, it means we already deviate from Openbox and it sounds sensible to use `<prompt text="foo" />` syntax.

> By the way, have you explore the stdin functionality for future config-error-logging. Try echo "foo\nbar" | labnag -l -m baz

I confirmed it worked, but I wonder why we need it. Can't we just pass the error messages via command line? Or are you planning to launch `labnag` as a daemon and let it show error messages by receiving messages from labwc via pipe?

> Assume you mean -m instead of the lonely -?

Yes.

> Do you mean that we should support Z and B but remove support for z and b? I'm thinking it's better to just pass things to execlp("sh", "sh", "-c", button->action, NULL); rather than invoking terminals.

Yes, I meant that we can support only `--button`/`--button-dismiss` and make them run a command without terminal emulator, just like what `--button-no-terminal`/`--button-dismiss-no-terminal` are currently doing.

> Do you mean to remove support for -s and simply use -z foo : instead?

Yes.

> One thing that doesn't work with this approach is having multiple actions with different option values.

Yeah, it's a trade-off between flexibility and simplicity. I guess you mean something like having `Yes`/`No` options for some actions but also having `Exit`/`Don't Exit` options for `<action name="Exit">`. But I think we can prioritize simplicity here since `Yes`/`No` options are sufficient for most cases.

johanmalm commented:
v3:
- Added documentation
- Squashed a few commits and tidied up commit messages

This is now ready for review. I'll leave PR at draft until after `0.9.0`

tokyo4j commented on src/action.c:
```diff
@@ -773,6 +778,112 @@ view_for_action(struct view *activator, struct server *server,
+
+		wlr_log(WLR_INFO, "Found pending prompt for exit code %d", exit_code);
+		struct wl_list *actions = NULL;
+		/*
+		 * TODO: figure out how to have more than 2 choices,
+		 *       maybe by using branch names like cond_0, cond_1
+		 *       and so on and falling back to 'then' and 'else'
+		 *       if not found?
+		 */
```
This comment should be removed as we agreed that this is out of scope.

Consolatis commented:
> I wonder which of `<prompt>` or `<query>` should be prioritized. In my patch above, when there are both `<prompt>` and `<query>` in a `If` action, it always spawns a window dialog, and if "yes" is selected, `<query>` is checked. But this PR seems to prioritize `<prompt>`.

Well, technically it should depend on the order of the queries / prompts. But in practice this seems way too complicated to implement. I don't have a strong opinion about all prompts first or all queries first.

johanmalm commented:
Rebased to resolve merge conflicts.

tokyo4j commented:
> How do we best implement? Shall I squash all the [fixup] comments - and then you just push to this branch?

I opened #2960. Could you review & merge it first? Then you can rebase your branch on master.

johanmalm commented:
> Hm, I think my suggested approach was actually not the best.
> 
> I was tempted to allow `<prompt>` inside `ForEach` action after reading the FIXME comment, but now I don't think it's actually demanded. When a user wants to combine `ForEach` action and `<prompt>`, he can just use nested actions. I think it's more cleaner to not share `run_action()` for `action_check_prompt_result()` and `actions_run()`.
> 
> I updated [my branch](https://github.com/johanmalm/labwc/compare/nag...tokyo4j:labwc:nag-foreach). I also restored the behavior of not running child actions in `If` action when no window is focused, which I changed in #2960. Sorry for my bad steering!

Okay. Shall I rebase my nag branch and then we can apply you new commit on top of it?

tokyo4j commented:
> Okay. Shall I rebase my nag branch and then we can apply your new commit on top of it?

Yes, or you can write your own commit and mention my name as `Co-authored-by`.

johanmalm commented:
> OK, I can help that. I will create `nag` branch to my repo so that you can use it.

Thanks. That helped a lot.

Okay, I think we're finally ready for review with a view to merge.

tokyo4j commented on clients/labnag.c:
```diff
@@ -0,0 +1,1649 @@
+		switch (c) {
+		case 'B': /* Button */
+		case 'Z': /* Button (Dismiss) */
+			if (nag) {
```
Can `nag` be NULL?

tokyo4j commented:
When I run labnag with `labnag -m hello -Z close :` and press the close button with my touchpad, labnag is not closed immediately. It's because `poll()` is called even after `nag->run_display` is set to `false` in `wl_display_dispatch_pending()`. The patch below fixes this:

```diff
diff --git a/clients/labnag.c b/clients/labnag.c
index f0b17f77..0d2b32a2 100644
--- a/clients/labnag.c
+++ b/clients/labnag.c
@@ -1192,7 +1192,7 @@ nag_run(struct nag *nag)
 			.events = POLLIN,
 		}
 	};
-	while (nag->run_display) {
+	while (true) {
 		while (wl_display_prepare_read(nag->display) != 0) {
 			wl_display_dispatch_pending(nag->display);
 		}
@@ -1202,6 +1202,10 @@ nag_run(struct nag *nag)
 			break;
 		}
 
+		if (!nag->run_display) {
+			break;
+		}
+
 		poll(fds, 2, -1);
 		if (fds[0].revents & POLLIN) {
 			if (timer_fd >= 0 && nag->details.close_timeout_cancel) {
```

I'll continue reviewing.

tokyo4j commented:
I find it a bit strange that `labnag -m hello -B option1 : -B option2 :` places "option1" button at right and "option2" button at left. Can we inverse the button order?

tokyo4j commented on clients/labnag.c:
```diff
@@ -0,0 +1,1649 @@
+		case 'B': /* Button */
+		case 'Z': /* Button (Dismiss) */
+			if (nag) {
+				if (optind >= argc) {
+					fprintf(stderr, "Missing action for button %s\n", optarg);
+					return LAB_EXIT_FAILURE;
+				}
```
Rather than requiring the action argument, can we just make it optional so that it can also function as swaynag's `-s` which was removed in response to my comment (https://github.com/labwc/labwc/pull/2699#issuecomment-2875286896)?

tokyo4j commented on clients/labnag.c:
```diff
@@ -0,0 +1,1649 @@
+}
+
+static void
+button_execute(struct nag *nag,
+		struct button *button)
```
Nit: unnecessary indentation. Same goes for other functions.

tokyo4j commented on clients/labnag.c:
```diff
@@ -0,0 +1,1649 @@
+	}
+
+	int border = nag->conf->bar_border_thickness;
+	if (max_height > nag->height) {
+		max_height += border;
+	}
```
What is this `if` branch for? This is causing 1 more call to `wl_display_roundtrip()` in `render_frame()` when the detail is being opened.

johanmalm replied:
Yeah, I think the logic could be improved, simplified, re-factored... but I have not touched this in the basis that it's stood the test of time and I feel it's safer at this stage rather than embarking on an overhaul

johanmalm commented:
Thanks for the review. Very thorough. I'll work through the comments, but might take a few days.

johanmalm commented:
@tokyo4j Thanks for your `nag` branch. I've cherry-picked those four commits.

Also updated `labnag(1)` to reflect the change relating to the `-B/-Z` options (optional action argument and the order the buttons are shown).

Have squashed, but updated the commit message with a `v3` section to reflect this.

tokyo4j commented:
I hid my comments that are now resolved. Remaining are:

1. Remove `close_timeout_cancel`?
2. Remove unnecessary indentations
3. Remove `if (max_height > nag->height)` if it's not needed
4. Limit the length of detail messages to prevent overloading CPU

I think (1) and (2) are easily addressed. I'm not sure about (3) and (4). If they are not easy, let's add a `TODO` comment and address them later.

LGTM other than that. Actually I don't understand all the logic of `labnag.c`, but everything seems to be working well in my testing.

johanmalm commented on clients/labnag.c:
```diff
@@ -0,0 +1,1649 @@
+		char *message;
+		char *details_text;
+		int close_timeout;
+		bool close_timeout_cancel;
```
Thanks. Have removed redundant variable `close_timeout_cancel`
